llumnix/global_scheduler/migration_scheduler.py

```diff
@@ -82,31 +135,29 @@ def __init__(self,
 
     @abstractmethod
     def pair_migration(self,
-                      sorted_instance_infos: List[InstanceInfo]
-                      ) -> List[Tuple[str, str]]:
+                       sorted_src_instance_infos: List[InstanceInfo],
+                       sorted_dst_instance_infos: List[InstanceInfo],
+                       pre_migration: bool,
+                       ) -> List[Tuple[str, str]]:
         raise NotImplementedError
 
 class Balanced(PairMigrationPolicy):
     def pair_migration(self,
-                       sorted_instance_infos: List[InstanceInfo]
+                       sorted_src_instance_infos: List[InstanceInfo],
+                       sorted_dst_instance_infos: List[InstanceInfo],
+                       pre_migration: bool,
                        ) -> List[Tuple[str, str]]:
-        # migrate in instances
-        migrate_in_instance_infos = [i for i in sorted_instance_infos
-                               if i.num_killed_requests == 0 and i.instance_load_migrate < self.migrate_out_load_threshold]
-        # migrate out instances
-        migrate_out_instance_infos = [i for i in reversed(sorted_instance_infos)
-                                if i.num_killed_requests > 0 or i.instance_load_migrate > self.migrate_out_load_threshold]
         migrate_instance_pairs = []
-        for i in range(min(len(migrate_in_instance_infos), len(migrate_out_instance_infos))):
-            load_diff_before_mig = migrate_out_instance_infos[i].instance_load_migrate - migrate_in_instance_infos[i].instance_load_migrate
-            left_load_after_mig = self._compute_instance_load_after_migrate(migrate_in_instance_infos[i], is_migrate_in=True)
-            right_load_after_mig = self._compute_instance_load_after_migrate(migrate_out_instance_infos[i], is_migrate_in=False)
+        for i in range(min(len(sorted_src_instance_infos), len(sorted_dst_instance_infos))):
+            load_diff_before_mig = sorted_src_instance_infos[i].instance_load_migrate - sorted_dst_instance_infos[i].instance_load_migrate
+            left_load_after_mig = self._compute_instance_load_after_migrate(sorted_dst_instance_infos[i], is_migrate_in=True)
+            right_load_after_mig = self._compute_instance_load_after_migrate(sorted_src_instance_infos[i], is_migrate_in=False)
             # Add some constrains to reduce unnecessary migrations
             if left_load_after_mig > self.migrate_out_load_threshold:
```

Comment:
> is there something wrong with this? migrate_out_load_threshold refers to "Migrate out instance load threshold". left_load_after_mig is the load of the dst.

If the load of the dst exceeds the threshold after migration, it will also be migrated out in the next pair migration, so it’s better to add this restriction.